wlauto/common/android/workload.py

```diff
@@ -289,14 +289,21 @@ def _grant_requested_permissions(self):
         for line in lines:
             if "android.permission." in line:
                 permissions.append(line.split(":")[0].strip())
-            else:
+            # Matching either of these means the end of requested permissions section
+            elif "install permissions:" in line or "runtime permissions:" in line:
                 break
 
-        for permission in permissions:
+        for permission in set(permissions):
             # "Normal" Permisions are automatically granted and cannot be changed
             permission_name = permission.rsplit('.', 1)[1]
             if permission_name not in ANDROID_NORMAL_PERMISSIONS:
-                self.device.execute("pm grant {} {}".format(self.package, permission))
+                # On some API 23+ devices, this may fail with a SecurityException
+                # on previously granted permissions. In that case, just skip as it
+                # is not fatal to the workload execution
+                try:
+                    self.device.execute("pm grant {} {}".format(self.package, permission))
+                except DeviceError as e:
```

Comment:
You should check that the error message contains what you would expect (e.g. words "is already granted" or whatever is actually in the message), and only debug-log it then, otherwise re-raise.